lib/space_saving.c

```diff
@@ -0,0 +1,795 @@
+			RTE_LOG(ERR, HASH, "Unknown network layer protocol\n");
+			ret = -1;
+		break;
+	}
+	return ret;
+}
+
+/* Add data to a counter bucket */
+static int
+add_bucket_entry(struct counter_bucket *ct_bkt, struct ip_flow *flow,
+			struct bucket_data *bkt_data)
+{
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
+			ret = rte_hash_add_key_data(ct_bkt->bkt.bkt_ip4,
+							flow, bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be added into "
+					"the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_IPv6:
+			ret = rte_hash_add_key_data(ct_bkt->bkt.bkt_ip6,
+							flow, bkt_data);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be added into " 
+					"the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_ARP:
+			ret = -1;
+			break;
+
+		default:
+			RTE_LOG(ERR, HASH,"Unknown network layer protocol!\n");
+			ret = -1;
+			break;
+	}
+	return ret;
+}
+
+/* Deletes data from a counter bucket */
+static int
+delete_bucket_entry(struct counter_bucket *ct_bkt, struct ip_flow *flow)
+{
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
+			ret = rte_hash_del_key(ct_bkt->bkt.bkt_ip4, flow);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be deleted "
+					"from the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_IPv6:
+			ret = rte_hash_del_key(ct_bkt->bkt.bkt_ip6, flow);
+			if(ret < 0) {
+				RTE_LOG(ERR, HASH, "Data cannot be deleted "
+					"from the bucket\n");
+				ret = -1;
+			}
+			break;
+
+		case ETHER_TYPE_ARP:
+			ret = -1;
+			break;
+
+		default:
+			RTE_LOG(ERR, HASH,"Unknown network layer protocol!\n");
+			ret = -1;
+			break;
+	}
+	return ret;
+}
+
+/* Set up parameters for a counter table */
+static int
+setup_counter_params(unsigned int socket_id, int identifier, int ht_size,
+			int ip_ver, struct rte_hash_parameters *ct_hash_params)
+{
+
+	int ret;
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+		sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	ct_hash_params->name = NULL;
+	ct_hash_params->entries = ht_size;
+	ct_hash_params->key_len = key_len;
+	ct_hash_params->hash_func = DEFAULT_HASH_FUNC;
+	ct_hash_params->hash_func_init_val = 0;
+	ct_hash_params->reserved = 0;
+
+	char ct_name[64];
+
+	ret = snprintf(ct_name, sizeof(ct_name), "counter_hash_%d", identifier);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(ct_name));
+
+	ct_hash_params->name = ct_name;
+	ct_hash_params->socket_id = socket_id;
+
+	return ret;
+}
+
+/* Create a counter table */
+static int
+create_counter_table(struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int ret;
+	unsigned int socket_id;
+	struct rte_hash_parameters ct_hash_params;
+	struct counter_table *ct_table = malloc(sizeof(*ct_table));
+
+	socket_id = rte_lcore_to_socket_id(gk_conf->lcores[0]);
+
+	/* 
+	 * TODO : How to get max_size for counter table ?
+	 * Currently it has been set to 1000000 but final values to be updated 
+	 * after testing.
+	 */
+
+	if(ipv4_if_configured(iface)) {
```

Comment:
Add a space after the `if` and before the parenthesis. Look through the rest of the code for similar cases.